update-format-crau/src/bin/test.rs

```diff
@@ -79,12 +83,91 @@ fn main() -> Result<(), Box<dyn Error>> {
         _ => None,
     };
 
-    let signatures = match signatures_bytes {
-        Some(ref bytes) => Some(proto::Signatures::parse_from_bytes(bytes)?),
-        None => None,
+    Ok(signatures_bytes.unwrap())
+}
+
+// parse_signature_data takes SigBuffer and public key file path.
+// Return only actual data, without version and special fields.
+fn parse_signature_data(sigbytes: &[u8], pubkeyfile: &str) -> Result<Box<[u8]>, Box<dyn Error>> {
+    // Signatures has a container of the fields, i.e. version, data, and
+    // special fields.
+    let sigmessage = proto::Signatures::parse_from_bytes(sigbytes)?;
+    let siglen = sigbytes.len();
+
+    let mut sigbs: Box<[u8]> = vec![0; siglen].into_boxed_slice();
+    let mut sigfound: bool = false;
+
+    // sigmessages.signatures[] has a single element in case of dev update payloads,
+    // while it could have multiple elements in case of production update payloads.
+    // For now we assume only dev update payloads are supported.
+    // Return the first valid signature, iterate into the next slot if error.
+    for sigx in &sigmessage.signatures {
+        match verify_sig_pubkey(sigx, pubkeyfile) {
+            Ok(sigxvec) => {
+                sigbs = sigxvec.into_boxed_slice();
+                sigfound = true;
+                break;
+            }
+            Err(_) => {
+                continue;
+            }
+        };
+    }
+
+    if !sigfound {
+        return Err("No signature found".into());
+    }
+
+    Ok(sigbs)
+}
+
+// Verify signature with public key
+fn verify_sig_pubkey(sig: &Signature, pubkeyfile: &str) -> Result<Vec<u8>, Box<dyn Error>> {
+    // The signature version is actually a numeration of the present signatures,
+    // with the index starting at 2 if only one signature is present.
+    // The Flatcar dev payload has only one signature but
+    // the production payload has two from which only one is valid.
+    // So, we see only "version 2" for dev payloads , and "version 1" and "version 2"
+    // in case of production update payloads. However, we do not explicitly check
+    // for a signature version, as the number could differ in some cases.
+    debug!("supported signature version: {:?}", sig.version());
+    let sigvec = if sig.has_data() {
+        debug!("data: {:?}", sig.data());
+        sig.data().into()
+    } else {
+        return Err("no signature found".into());
     };
```

Comment:
That sounds like a good idea.

However, `sig.data()` does not return `Option`. So the above fails to compile:

```
error[E0599]: no method named `ok_or_else` found for reference `&[u8]` in the current scope                   
   --> update-format-crau/src/bin/test.rs:133:29                                                              
    |                                                                                                         
133 |     let sigvec = sig.data().ok_or_else(|| return Err("no signature found".into()));                     
    |                             ^^^^^^^^^^ method not found in `&[u8]`                                      
                                                                                                              
For more information about this error, try `rustc --explain E0599`.                                           
error: could not compile `update-format-crau` (bin "test") due to previous error  
```

Reply:
oh good catch, i thought there was something i was missing.

looks like the `data` field is public though, so maybe
```rust
let sigvec = sig.data
    .ok_or_else(|| Err("no signature found".into()))?;
```
works? (`sig.data` instead of `sig.data()`)

Reply:
Yes, that works in a way. Thanks.
But compile error does not simply go away due to a return type issue. Will comment below.